### PR TITLE
export corelogic data for forms, photos, videos

### PR DESCRIPTION
**Pull request by nelsoneab4**
### Link to Jira Card
[ENG-538](https://flyreel.atlassian.net/browse/ENG-538?atlOrigin=eyJpIjoiNWFjZmE2ZDM2YzA4NDllM2FiYjJmODMxOTZmNDYzNTkiLCJwIjoiaiJ9)

### Description
Added cloud function to export CoreLogic inspection data. Function exports form data, photos and videos. Photos and videos are sent one at a time and a new object is added to inspection metadata to check for sending duplicate photos / videos.

### CODE REVIEW

File: src/Handlers/exportInspection/exportInspection.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    inspectionId = inspection._id;
+    const externalId = inspection.meta?.external_id;
+
+    if (!externalId) {
+      throw new Error(`Missing required field external_id`);
+    }
+
+    const coreLogicToken = await getToken();
+    const inspectionDetails = await axios.get(
+      //probably some headers too
```
Comment (rosully) [inline src/Handlers/exportInspection/exportInspection.ts:35]: Is this comment still needed?

File: src/Handlers/exportInspection/exportInspection.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      {
+        data: {
+          InspectionId: inspectionId,
+          UniqueId: externalId,
+          formData,
+        },
+      }
+    );
+
+    for (const [key, values] of Object.entries(photos)) {
```
Comment (YongDeng22) [inline src/Handlers/exportInspection/exportInspection.ts:67]: photos could be array

File: src/Handlers/exportInspection/exportInspection.ts [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        const form = new FormData();
+        const formData = {
+          InspectionId: inspectionId,
+          UniqueId: externalId,
+          [key]: value,
+        };
+        Object.keys(formData).forEach((key) => {
+          form.append(key, form[key]);
+        });
+        axios.post(
```
Comment (YongDeng22) [inline src/Handlers/exportInspection/exportInspection.ts:78]: Missing await

File: src/Handlers/exportInspection/exportInspection.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+          `${corelogicApiUrl}/api/digitalhub/v1/Action/State?InspectionId=${externalId}&UniqueId=${inspectionId}`,
```
Comment (YongDeng22) [inline src/Handlers/exportInspection/exportInspection.ts:79]: The path for photo upload is "/api/digitalhub/v1/Photo/Upload"

File: src/Handlers/exportInspection/exportInspection.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+          UniqueId: externalId,
+          formData,
+        },
+      }
+    );
+
+    for (const [key, values] of Object.entries(photos)) {
+      (values as any).forEach((value) => {
+        const form = new FormData();
+        const formData = {
```
Comment (YongDeng22) [inline src/Handlers/exportInspection/exportInspection.ts:70]: Missing "Label". It is a required field.

File: src/Handlers/exportInspection/exportInspection.ts [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+          {
+            headers: {
+              Authorization: `Bearer ${coreLogicToken}`,
+              "Content-Type": "application/json",
+              "api-key": apiKey,
+              "api-companyid": apiCompanyId,
+            },
+          },
+          {
+            data: {
```
Comment (YongDeng22) [inline src/Handlers/exportInspection/exportInspection.ts:89]: body should be the second parameter of axios.post

File: src/Handlers/exportInspection/exportInspection.ts [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+      const formData = {
+        InspectionId: inspectionId,
+        UniqueId: externalId,
+        [key]: value,
+      };
+      Object.keys(formData).forEach((key) => {
+        form.append(key, form[key]);
+      });
+      await axios.post(
+        `${corelogicApiUrl}/api/digitalhub/v1/Action/State?InspectionId=${externalId}&UniqueId=${inspectionId}`,
```
Comment (YongDeng22) [inline src/Handlers/exportInspection/exportInspection.ts:108]: The path for video upload is "/api/digitalhub/v1/Video/Upload"

File: src/Handlers/exportInspection/exportInspection.ts [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        );
+      });
+    }
+
+    for (const [key, value] of Object.entries(videos)) {
+      const form = new FormData();
+      const formData = {
+        InspectionId: inspectionId,
+        UniqueId: externalId,
+        [key]: value,
```
Comment (YongDeng22) [inline src/Handlers/exportInspection/exportInspection.ts:102]: We may need to check with CoreLogic for one of the keys. In their notes, it says there should be field called "File" while in the example it shows as "VideoFile".

File: src/Handlers/exportInspection/exportInspection.ts [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        {
+          headers: {
+            Authorization: `Bearer ${coreLogicToken}`,
+            "Content-Type": "application/json",
+            "api-key": apiKey,
+            "api-companyid": apiCompanyId,
+          },
+        },
+        {
+          data: {
```
Comment (YongDeng22) [inline src/Handlers/exportInspection/exportInspection.ts:118]: form should be the second parameter:
`axios.post('https://example.com', form, { headers: form.getHeaders() })`

File: src/Handlers/exportInspection/exportInspection.ts [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+    await logEvent({
+      inspection: inspectionId,
+      event: "inspection_data_sent_to_corelogic_notification",
+      meta: { external_id: externalId },
+    });
+
+    res.status(200).send(response);
+  } catch (error) {
+    console.error(
```
Comment (YongDeng22) [inline src/Handlers/exportInspection/exportInspection.ts:145]: Should an error code be sent back? We don't have a retry mechanism though. We can discuss further.

File: src/Handlers/exportInspection/exportHelpers.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
+import emojiRegex from "emoji-regex";
+
+const emojis = emojiRegex();
+const punctuation = /[\u2000-\u206F\u2E00-\u2E7F\\'!"#$%&()*+,\-.\/:;<=>?@\[\]^_`{|}~]/g;
+const spaces = /\s+/g;
+
+export const transformFormData = (inspection: any): any => {
+  const conversation = inspection.conversation;
+  const fieldData = [] as any;
```
Comment (YongDeng22) [inline src/Handlers/exportInspection/exportHelpers.ts:9]: Should the type be any[] to be more explicit?

File: src/Handlers/exportInspection/exportInspection.ts [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+          "Content-Type": "application/json",
+          "api-key": apiKey,
+          "api-companyid": apiCompanyId,
+        },
+      }
+    );
+
+    await logEvent({
+      inspection: inspectionId,
+      event: "inspection_data_sent_to_corelogic_notification",
```
Comment (YongDeng22) [inline src/Handlers/exportInspection/exportInspection.ts:139]: This event is not a notification. Could it be called "inspection_data_sent_to_corelogic"?

File: src/Handlers/exportInspection/exportInspection.ts (hunk 1 above)
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportInspection.ts:35]: Removed comment

File: src/Handlers/exportInspection/exportHelpers.ts [hunk 13]
```diff
@@ -0,0 +1,113 @@
+import emojiRegex from "emoji-regex";
+
+const emojis = emojiRegex();
+const punctuation = /[\u2000-\u206F\u2E00-\u2E7F\\'!"#$%&()*+,\-.\/:;<=>?@\[\]^_`{|}~]/g;
```
[the remark this answers is not in the document]
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportHelpers.ts:4]: Yep, changed to only allow alpha-numeric.

File: src/Handlers/exportInspection/exportHelpers.ts [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+  messages.forEach((message: any) => {
+    if (
+      message.show_in_dashboard &&
+      message.completed &&
+      message.type !== "photo" &&
+      message.type !== "video"
+    ) {
+      fieldData.push({
+        Name: message.name.replace(punctuation, "").replace(spaces, ""),
```
[the remark this answers is not in the document]
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportHelpers.ts:28]: Changed name format.

File: src/Handlers/exportInspection/exportInspection.ts [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+        },
+      }
+    );
+
+    const formData = transformFormData(inspectionDetails);
+    const photos = transformPhotos(inspectionDetails);
+    const videos = transformVideos(inspectionDetails);
+
+    await axios.post(
+      `${corelogicApiUrl}/api/digitalhub/v1/Action/State?InspectionId=${externalId}&UniqueId=${inspectionId}`,
```
[the remark this answers is not in the document]
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportInspection.ts:49]: path updated.

File: src/Handlers/exportInspection/exportInspection.ts [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+      {
+        headers: {
+          Authorization: `Bearer ${coreLogicToken}`,
+          "Content-Type": "application/json",
+          "api-key": apiKey,
+          "api-companyid": apiCompanyId,
+        },
+      },
+      {
```
[the remark this answers is not in the document]
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportInspection.ts:58]: Changed data to be second parameter. Form data follows the CoreLogic form data format and is stored as the formData object.

File: src/Handlers/exportInspection/exportInspection.ts (hunk 4 above)
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportInspection.ts:79]: path updated.

File: src/Handlers/exportInspection/exportInspection.ts (hunk 3 above)
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportInspection.ts:78]: added async await

File: src/Handlers/exportInspection/exportInspection.ts (hunk 6 above)
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportInspection.ts:89]: body changed to second parameter.

File: src/Handlers/exportInspection/exportInspection.ts (hunk 9 above)
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportInspection.ts:118]: changed to be second parameter.

File: src/Handlers/exportInspection/exportHelpers.ts (hunk 11 above)
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportHelpers.ts:9]: changed to be more explicit.

File: src/Handlers/exportInspection/exportInspection.ts (hunk 12 above)
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportInspection.ts:139]: changed name to remove notification.

File: src/Handlers/exportInspection/exportInspection.ts (hunk 7 above)
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportInspection.ts:108]: updated path.

File: src/Handlers/exportInspection/exportHelpers.ts [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+  messages.forEach((message: any) => {
+    const messageName = message.name
+      .replace(punctuation, "")
+      .replace(spaces, "");
+    if (
+      message.show_in_dashboard &&
+      message.completed &&
+      message.type === "photo" &&
+      !photos[messageName]
```
[the remark this answers is not in the document]
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportHelpers.ts:65]: changed photos to object type

File: src/Handlers/exportInspection/exportHelpers.ts [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+    UniqueId: inspection.meta?.external_id,
+    Field: fieldData,
+  };
+
+  return formUpload;
+};
+
+export const transformPhotos = (inspection: any): any => {
+  const conversation = inspection.conversation;
+  const photos = inspection.meta?.exportedPhotos ?? [];
```
[the remark this answers is not in the document]
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportHelpers.ts:45]: changed photos to object type

File: src/Handlers/exportInspection/exportHelpers.ts [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+    }
+  });
+
+  inspection.meta.exportedPhotos = photos;
+  return photos;
+};
+
+export const transformVideos = (inspection: any): any => {
+  const conversation = inspection.conversation;
+  const videos = inspection.meta?.exportedVideos ?? [];
```
[the remark this answers is not in the document]
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportHelpers.ts:88]: changed videos to object type

File: src/Handlers/exportInspection/exportInspection.ts (hunk 2 above)
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportInspection.ts:67]: changed photos to object type where key is name and value is array of photos (handles case where multiple photos for one name exist).

File: src/Handlers/exportInspection/exportHelpers.ts [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+  });
+
+  messages.forEach((message: any) => {
+    if (
+      message.show_in_dashboard &&
+      message.completed &&
+      message.type === "video" &&
+      !videos[message.name]
+    ) {
+      videos[message.name] = message.answer;
```
[the remark this answers is not in the document]
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportHelpers.ts:107]: validating file types

File: src/Handlers/exportInspection/exportHelpers.ts [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+      photos[messageName] = [];
+      photos[messageName].push(message.answer);
+    } else if (
+      message.show_in_dashboard &&
+      message.completed &&
+      message.type === "video" &&
+      !photos[messageName]
+    ) {
+      photos[messageName] = [];
+      message.detections.forEach((detection: any) => {
```
[the remark this answers is not in the document]
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportHelpers.ts:76]: added validation of file types

File: src/Handlers/exportInspection/exportInspection.ts (hunk 8 above)
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportInspection.ts:102]: Resolved with CoreLogic updated docs

File: src/Handlers/exportInspection/exportInspection.ts (hunk 10 above)
Comment (nelsoneab4) [reply src/Handlers/exportInspection/exportInspection.ts:145]: Triggering slack notification for errors.

**Pull request closed without merging**